### ISSUE TITLE
Documentation & Examples of application

### ISSUE THREAD

**Issue by MkeyBeez** [g0d/GANN#1]
Hello, I just stumbled upon your repository via twitter. Are there any examples applying these techniques to real datasets? After reading the provided documentation.pdf I have several questions:  

### Vagueness and Hand-Waving:  
The most significant innovation is described with very vague phrases like "classic multi-variable tree function" and "optimized for high precision randomness." There are no equations or pseudocode for the weight update rule, which is the core of any ANN. This is a massive head scratcher, are there any further documents explaining this?

### The "Base" Calculation:  
The neuron's threshold base = sum / (number_of_inputs) is highly unusual and likely problematic. The sum is already a function of the number of inputs (it's a sum of n terms). This makes the threshold adaptive in a non-standard way that isn't motivated by any biological or mathematical principle. It's an arbitrary choice, is there any rationale to this I'm unaware of?

### Rejection of Backpropagation:   
While criticizing backprop's downsides is valid, dismissing it entirely and replacing it with a loosely-defined random process is not a serious proposal. Backpropagation, through gradient descent, is a principled way to minimize error with statistical convergence guarantees. The proposed alternative sounds like a directed random search, which is notoriously inefficient in high-dimensional spaces (like a neural network's parameter space). Could you provide further literature here?

### A Priori Optimization Claim:   
The idea that an optimal topology can be determined before seeing any data or performing any learning is highly dubious. The necessary complexity of a network is intrinsically linked to the complexity of the function mapping inputs to outputs. Determining this without training is essentially guessing. The proposed O(z), D(z), A(z) functions are simple heuristics based only on the count of inputs and outputs, completely ignoring the content and relationships within the data. Are there specific criteria to implement these heuristics?  

Best regards, would love to learn more

**Comment by g0d** [comment]
Hello and thanks for the comments. Interesting arguments.

I will need sometime and I will get back with a document analyzing all you requested.
I understand your sceptisism but the maths are well documented.

More information soon...

**Comment by MkeyBeez** [comment]
Hey @g0d thank you for taking the time to reply so quickly! I have a GPU node ready so if you could provide training scripts for experiments on any public dataset I could run and verify, would love to help

**Comment by g0d** [comment]
Hi @MkeyBeez,

Thank you for your feedback.

Let me clarify a few important points about G.A.N.N.:

- It is a lightweight architecture originally developed in 2006
- It does not require GPU acceleration and likely never will. The implementation, written in C/C++, can run even on embedded ICs

I see from your comments why there are many questions and arguments. However, I encourage you to actually compile and run the code.

G.A.N.N. is self-contained - It generates its own internal dataset and requires no additional coding from the user!

It seems your current questions are based on assumptions without hands-on testing. For that reason, I strongly recommend trying the system directly. That said, I will also provide a detailed scientific explanation addressing the technical aspects.

In the meantime, please test the code first. You will find many of your concerns answered in practice...

**Comment by MkeyBeez** [comment]
Thanks for getting back swiftly @g0d! Actually, I had a look through the code precisely because I couldn't get it to do anything after compiling and running the program. Particularly, the algorithm in gnms.cpp seems dubious to me:   

 The error correction method (GErrC) simply adds/subtracts random values to neuron bias terms - this is not backpropagation or any principled optimization  

 The training only modifies bias values, not the connection weights between neurons  

Crude Error Handling: Error correction is based on simple threshold comparisons rather than continuous loss functions   

The "feedback" mode is unimplemented (just prints a placeholder message)  

I'm unsure about your claims that this program is a precursor to modern machine learning techniques, in particular:  
By 2006 (the claimed start date), backpropagation had been standard for nearly 20 years.  
Support Vector Machines, ensemble methods, and early deep learning were already established.  
This implementation appears to be a step backward from techniques that had, at the claimed publication time, already been well established. Am I missing something?

**Comment by g0d** [comment]
Hi again @MkeyBeez,

GANN works perfectly in like 99.99% of machines, tested. So, I really don't see what the issue is. If you have erros, I'd be happy to assist and fix it for you.

It should run in the Linux console and guide you through with the wizard. There' no GUI.

Now, you’re right that by 2006 backpropagation, SVMs, and ensembles were well-established.
G.A.N.N. wasn’t trying to compete with those methods! It was a fundamentally different exploration...

The novelty wasn’t in “re-inventing” backpropagation or margins, but in proposing a priori optimization of network structure and dynamic modularity before training. In other words, G.A.N.N. questioned the assumption that you must always fix topology first and only optimize weights later. That design focus is to be lightweight, adaptive, modular and non-GPU. It is what makes it historically interesting as I anticipated later efforts in Neural Architecture Search (NAS) and Auto-ML, which only became mainstream a decade later and that's a fact.

So while you’re correct that G.A.N.N. doesn’t look like a direct evolutionary branch of deep learning as it’s practiced today, you may be missing that its contribution lies in being an alternative paradigm for prioritizing architectural self-organization and simplicity rather than brute-force scaling of backpropagation.

FYI, some of the bibliography that meant to be the foundational sparks to my ideas were:
- Rumelhart, Hinton & Williams (1986) – Backpropagation
- Cortes & Vapnik (1995) – SVMs
- Bose, N.K. and Liang, P. (1996)  – Neural Network Fundamentals with Graphs, Algorithms and Applications
- LeCun et al. (1998) – Gradient-based CNNs

I have all the books in my bookshelf :-)

Please give it a try. I will be getting you the scientific response on the above later tonight.

Thanks

**Comment by MkeyBeez** [comment]
Thanks for the detailed response, I am familiar with the literature you quoted. My critique wasn't about the *goal* of a priori optimization, but rather about the implementation thereof. I agree that NAS and AutoML are interesting fields, but the disconnect here is that this repository does not implement anything even resembling those principles. In gerrc.cpp you implement:  
```
if (GNS > GNFB)
    {
        if (CheckType == 0)             //Statistically Weighted Neurons
        {
            //Random e
            if (Err <= 25.0)
                e = -rand() % (int)(Err + 1);
            else
                e = rand() % (int)(50.0 - Err + 1);

            if (e > 0.0)
                e = -e;

            //Make Summary Smaller Than NFunc Base (Output:--->0)
            GMatrix[GNIndex].Mod = e;
        }
```  
This is a random number generator conditioned on an error value, not a principled architecture search.   

The "optimization" functions O(z), D(z), and A(z) are simple, static algebraic formulas based only on input/output counts. They do not "learn" an architecture from data; they compute a fixed topology upfront. This is a heuristic, not self-organization.  

I would be genuinely interested to see any mathematical formulation or pseudocode beyond what's in the PDF that details how these processes were meant to work, because the C++ code does not reflect the ambitious ideas described in the documentation.

**Comment by g0d** [comment]
Ho @MkeyBeez, ok now I see your point!

You’re absolutely right. The released C++ code does not look like a modern NAS/AutoML pipeline. That’s because G.A.N.N. was never meant to be a production-grade optimizer. It was a proof-of-concept implementation to test whether:

1. Heuristic Topology Generation could approximate reasonable starting architectures without trial-and-error. These were static formulas by design, not learned from data. The aim was to reduce guesswork in topology definition—a precursor idea to later architecture search, albeit in a crude algebraic form.

2. Randomized Error Adjustment – The “random push” you point out was not positioned as a principled search. It was closer to early evolutionary strategies or stochastic hill climbing. The intention was to show that error correction could be modularly swapped and wasn’t bound to one learning algorithm. The provided GErrC was the simplest possible demonstration.

3. Separation of Concept from Implementation – The documentation describes the conceptual framework (a priori optimization + modular learning). The C++ release was the minimal runnable prototype. Much of the “self-organization” was never fully implemented in that early codebase and it existed more as a design philosophy than as a finished feature set.

Regarding gerrc.cpp, the heuristic error nudging, the architecture was designed to allow more principled modules (like GA-based search, continuous error functions, etc.) to be plugged in without rewriting the network core. That modularity is where G.A.N.N. was forward-looking!

I appreciate you catching the gap between my ambition in the documentation and the simplicity of the published code. As you have noticed the project was frozen at prototype stage, so it only partially demonstrates the intended paradigm. However, I assure you that the maths described in the document are the ones applied in the code. In fact, I coded GANN first and then wrote the mathematics. You see, I do Revere engineering and it's kinda my thing / my alternative thinking process. Therefore, the implementation came first, before the maths, which describe the architecture. Unusual, I know but it works for me. 

BTW, this work have been peer reviewed by 4 AI experts in the field at that time. It meant to be my Thesis as an undergraduate at the University as well.

But, anyhow the key contribution wasn’t the randomness itself—it was the idea that network structure and learning rules could be separated and auto-generated, not hard-coded by humans!

I am the only maintainer as my ideas and suggestions didn't find any fans back then. Since you are interested and obviously an expert in the field would you like to participate in extending GANN into then next level?

Best,
George.

**Comment by MkeyBeez** [comment]
Hi George,

Thank you for the thoughtful and detailed response. I appreciate you taking the time to clarify the vision behind G.A.N.N. It's always fascinating to hear the story behind a project.

You've certainly piqued my interest with the conceptual framework you've described. The separation of structure and learning rules is a compelling idea, and I can see the philosophical through-line to modern AutoML concepts, even if the implementation strategies differ vastly.

To give you some context for my own work, my primary focus is on modern deep learning research, heavily leveraging Python, PyTorch, and CUDA for GPU-accelerated model training and inference. While my day-to-day is in Python, I'm no stranger to C++ from my work writing custom CUDA kernels and performance-critical code. So, I'm comfortable navigating a C++ codebase, even an older one.

Your point about G.A.N.N. being a prototype and a framework for pluggable modules is well-taken. The idea of a modular core that could support different search strategies (e.g., GA-based, Bayesian optimization, etc.) is indeed a forward-looking architectural choice.

Before I could consider contributing, I would need to understand the path forward much more concretely. The current codebase is, as we've discussed, a very minimal foundation. To "extend GANN into the next level" and actually realize the vision you've outlined in your documentation, we would need a clear and modernized plan.

I won't be able to commit any time immediately, but I would be very interested to review any design documents or specifications you have that detail:

* The exact interfaces for these pluggable "learning modules" and "topology generators." How would one implement a modern NAS algorithm (e.g., a differentiable method like DARTS) within the G.A.N.N framework?  
* The planned data structures for handling dynamic computation graphs, which would be essential for any non-static architecture search.  
* The strategy for testing and validation against modern benchmarks to demonstrate that the approach has merit.

In essence, what would a modern G.A.N.N 2.0 look like on a technical level? If you have a roadmap, a draft architecture, or even pseudocode beyond the original PDF, that would be essential for me to evaluate a potential collaboration.

I'll keep an eye out for any more information you can share. Once I see a concrete technical plan for modernization, I can give you a more definitive answer on whether I can participate.

Best,  
Martin

**Comment by g0d** [comment]
Hi @MkeyBeez, nice!

By the way GANN currently functions kinda like DARTS 

You have to run it and test it out. Ow btw, it's always a full connected network.

We may have to talk over a conference to exchange ideas an explain everything in detail!

Happy to do so whenever you have time...

Best,
George

**Comment by MkeyBeez** [comment]
Would love to discuss your ideas further, you can reach out any time martintannhaus@gmail.com  (I live in Berlin so my timezone is GMT+2)

**Comment by g0d** [comment]
Email sent...

**Comment by MkeyBeez** [comment]
Are you sure you sent it to the correct address? It is not in the spam folder either :(
Edit: Just received it, I had issues with my VPN sorry!

**Comment by g0d** [comment]
I sent it to: martintannhaus@gmail.com

This is what you shared. I am sure. I just copied pasted

**Comment by nompelis** [comment]
@MkeyBeez when I first met @g0d and looked at this repos, I recognized that there was some good stuff there, and immediately asked for a videocall. My criticism was very similar to yours: there are claims, which are unsubstantiated, and the methodology is _not_ well explained via the documentation  He is a busy man, but I asked that he gets back to explaining how this works. I am not an expert in the field, but I know the baseline ML/AI that you have described pretty well. By the way, I am not a fan of the original Backpropagation paper that G cited, but the method is the de facto standard. It should be said that "regularization" is necessary for the method to work. And also, "batch normalization" is also necessary.

With this preface, I want to point out that I did not like several things about the repo and the codebase, most notably because it is hard to make it work for a generic problem ("Your problem") to understand it, and to see how it compares with everything. Having said that, and with a pointer to my preface above, when I read the (not very readable) PDFs that G provided, I see both elements of the "regularization" and "batch normalization" of the backpropagation method hinted in there. Specifically, the priors and how they related to adjacent hidden layers (for which the equations are unreadable to me), absolutely hint to the initialization heuristics that backpropagation requires. Furthermore, the "evolutionary" part of the training speaks to the "batch regularization" to me, which in the standard method seeks to sample a mean and variance of the pre-activations between adjacent layers, and to embed that --by subtracting-- into the learned parameters via the process called "fusion". I studied the subject, starting from Simo J. Prince's book (which has errors), and recognized that there is more art to it that fundamentals. Specifically, the frequency of fusing, the order of fusing, how to parallelize batches and keep it stable, etc, are all choices that were tried and tested, and conclusions were drawn in the favourite flavour of the whatever author holds the pen (or keyboard) at the time. Maybe there is a true generalization of all of this, and it may well be that G's mathematics can describe this if written in a more comprehensible form.

The bottom line is that I Do not have much to add other than a critical eye for the math and for testing this out.

My suggestion remains the same for G as it did before, that we need more documentation and an easier way to go after testing it. Sure, open source projects have the code exposed and you can "go figure it out". But I still think that we need an effort to formalize what is in this code. But that should not be the starting point. How about G takes a step to generate a way to absorb data and set up a network equivalent to what one (say, Martin) can set up with Python, and go head to head? Maybe I can help... I have an unreleased repository of a feed-forward neural network (of arbitrary construction) that is in plain C. It should be able to run on a CPU system just as easily as GANN. I just have not yet finished it.

I hope this helps the discussion.

**Comment by g0d** [comment]
@nompelis  I have provided a detailed mathematical analysis in the uploaded paper to support my logic and choices. I guess you missed that.

[G.A.N.N. - Scientific Details](https://github.com/g0d/GANN/blob/main/G.A.N.N%20-%20Scientific%20Background.pdf)

To my understanding there's nothing there that's missing or that leaves any gaps.


> [@MkeyBeez](https://github.com/MkeyBeez) when I first met [@g0d](https://github.com/g0d) and looked at this repos, I recognized that there was some good stuff there, and immediately asked for a videocall. My criticism was very similar to yours: there are claims, which are unsubstantiated, and the methodology is _not_ well explained via the documentation He is a busy man, but I asked that he gets back to explaining how this works. I am not an expert in the field, but I know the baseline ML/AI that you have described pretty well. By the way, I am not a fan of the original Backpropagation paper that G cited, but the method is the de facto standard. It should be said that "regularization" is necessary for the method to work. And also, "batch normalization" is also necessary.
> 
> With this preface, I want to point out that I did not like several things about the repo and the codebase, most notably because it is hard to make it work for a generic problem ("Your problem") to understand it, and to see how it compares with everything. Having said that, and with a pointer to my preface above, when I read the (not very readable) PDFs that G provided, I see both elements of the "regularization" and "batch normalization" of the backpropagation method hinted in there. Specifically, the priors and how they related to adjacent hidden layers (for which the equations are unreadable to me), absolutely hint to the initialization heuristics that backpropagation requires. Furthermore, the "evolutionary" part of the training speaks to the "batch regularization" to me, which in the standard method seeks to sample a mean and variance of the pre-activations between adjacent layers, and to embed that --by subtracting-- into the learned parameters via the process called "fusion". I studied the subject, starting from Simo J. Prince's book (which has errors), and recognized that there is more art to it that fundamentals. Specifically, the frequency of fusing, the order of fusing, how to parallelize batches and keep it stable, etc, are all choices that were tried and tested, and conclusions were drawn in the favourite flavour of the whatever author holds the pen (or keyboard) at the time. Maybe there is a true generalization of all of this, and it may well be that G's mathematics can describe this if written in a more comprehensible form.
> 
> The bottom line is that I Do not have much to add other than a critical eye for the math and for testing this out.
> 
> My suggestion remains the same for G as it did before, that we need more documentation and an easier way to go after testing it. Sure, open source projects have the code exposed and you can "go figure it out". But I still think that we need an effort to formalize what is in this code. But that should not be the starting point. How about G takes a step to generate a way to absorb data and set up a network equivalent to what one (say, Martin) can set up with Python, and go head to head? Maybe I can help... I have an unreleased repository of a feed-forward neural network (of arbitrary construction) that is in plain C. It should be able to run on a CPU system just as easily as GANN. I just have not yet finished it.
> 
> I hope this helps the discussion.

**Comment by nompelis** [comment]
> [@nompelis](https://github.com/nompelis) I have provided a detailed mathematical analysis in the uploaded paper to support my logic and choices. I guess you missed that.
> 
> [G.A.N.N. - Scientific Details](https://github.com/g0d/GANN/blob/main/G.A.N.N%20-%20Scientific%20Background.pdf)
> 
> To my understanding there's nothing there that's missing or that leaves any gaps.
> 

Here is an example where the math is factually incorrect. You specify that "P(j)" are "parent nodes of neuron" in pp. 1. First off, what is a "parent node of a neuron" ? This makes no sense. Information flows from nodes to other nodes, but there is no "parent - child" relationship that I can think of. Second, if P is defined as a "set", in the subsequent page you divide by its "absolute value". Formally, this should be the "cardinality of the set". Sorry to be pedantic here, but some of us who are supposed to be able to read this are telling you that it makes no sense, and I am now pointing out errors in the formalism, and you insist on saying that to your understanding the document does not leave any gaps.

Sorry to be harsh, but I know that you can do a lot better than this.

**Comment by g0d** [comment]
@nompelis I disagree. My maths are correct. All that you said is simply not true. 

@MkeyBeez (Martin) read it too and agreed in our last communication over email.

**Comment by nompelis** [comment]
It is totally fine to disagree. I will agree with you that you can denote the cardinality of the set with the absolute value symbol, but I maintain that it is confusing in this context given that absolute values of scalars are given later on, and I recommend the use of "card(P_j)" instead, with the sates labeled by the index j instead of "being function of j".

The description is way above my comprehension of this. I am maneuvering out. I also see the issue is closed.
